### PR TITLE
Introduce RestrictSUIDSGID= for disabling SUID/SGID file creation

### PR DESCRIPTION
**Pull request by poettering**
Let's lock things down further

### CODE REVIEW

File: src/shared/seccomp-util.c [hunk 1]
```diff
@@ -1795,3 +1799,62 @@ int seccomp_protect_hostname(void) {
...
+                if (r < 0)
+                        return r;
+
+                /* chmod, fchmod, fchmodat */
+
+                for (unsigned bit = 0; bit < 2; bit ++) {
+                        /* Block S_ISUID in the first iteration, S_ISGID in the second */
+                        mode_t m = bit == 0 ? S_ISUID : S_ISGID;
+
+                        r = seccomp_rule_add_exact(
```
Comment (ret2libc) [inline src/shared/seccomp-util.c:1829]: There are also open(), openat(), creat().

Comment (poettering) [reply src/shared/seccomp-util.c:1829]: you cannot create suid/sgid files with them

**Comment by topimiettinen** [comment]
Good idea and lgtm.

File: src/shared/seccomp-util.c (hunk 1 above)
Comment (ret2libc) [reply src/shared/seccomp-util.c:1829]:
```C
#define _GNU_SOURCE
#include <unistd.h>
#include <sys/syscall.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>

int main(int argc, char **argv) {
	syscall(SYS_open, "/tmp/mysuid", O_CREAT, S_IRWXU | S_IRWXG | S_IRWXO | S_ISUID);
	return 0;
}
```
This seems to create the suid/sgid file.

Comment (poettering) [reply src/shared/seccomp-util.c:1829]: hmm, true, but it has to remain empty, since on Linux the suid/sgid bits are dropped on write, iirc.

**Comment by poettering** [comment]
Force pushed a new version now. Only change: also block open/openat/creat/mkdir/mkdirat/mknod/mknodat when invoked with suid/sgid mode, as pointed out by @ret2libc. ptal!

**Comment by evverx** [comment]
I'm not sure what happened to networkd on CentOS CI but on Travis CI `systemd` built with ASan and UBSan run in a nspawn container failed to start all the basic services (that are run there):
```
systemd-journald.service        loaded failed failed Journal Service          
systemd-logind.service          loaded failed failed Login Service            
systemd-resolved.service        loaded failed failed Network Name Resolution  
systemd-journald-dev-log.socket loaded failed failed Journal Socket (/dev/log)
systemd-journald.socket         loaded failed failed Journal Socket           
```
with "Main process exited, code=exited, status=127/n/a". It could be that the new setting is incompatible with ASan and should be commented out in `create_asan_wrapper` along with `MemoryDeny*` and `SystemCall*`.

**Comment by evverx** [comment]
Regarding CentOS CI, I was wrong it wasn't `networkd`, turns out it timed out trying to boot up a machine (which is most likely related to this PR). cc @mrc0mmand.

**Comment by mrc0mmand** [comment]
It failed with the same error twice in a row, so it looks like it might be related to this PR. I'll give it one more try to have a *definitive* answer.

Failed jobs:
https://ci.centos.org/job/systemd-pr-build/5372/
https://ci.centos.org/job/systemd-pr-build/5370/

**Comment by evverx** [comment]
It's unrelated but having read https://seclists.org/oss-sec/2019/q1/179, I was wondering whether `libseccomp` should be bumped a notch in `meson.build` :-)

**Comment by mrc0mmand** [comment]
And yeah, there's a similar issue in CentOS CI as well:

```
[    4.788541] systemd[1]: systemd-udevd.service: Main process exited, code=exited, status=127/n/a
[    4.738384] systemd[1]: Child 281 (systemd-journal) died (code=exited, status=127/n/a)
...
```

**Comment by mrc0mmand** [comment]
Also, here's the full debug log: https://ci.centos.org/job/systemd-pr-build/5377/artifact/systemd-centos-ci/artifacts_boac7I/bootstrap-logs-upstream.I9T/sanity-boot-check.log

**Comment by mrc0mmand** [comment]
It looks like the `systemd-journald.service` fails because `systemd-journald-audit.socket` fails with `ConditionCapability=CAP_AUDIT_READ`, which, in turn, causes fail of `systemd-udevd-kernel.socket` and `systemd-udev.service` and that causes the fail in mounting the rootfs, thus the emergency shell.

**Comment by evverx** [comment]
Looks like on Ubuntu Xenial (even without ASan) `libpcre.so` can't be loaded:
```
13924 fstat(6</lib/x86_64-linux-gnu/libm-2.23.so>, {st_dev=makedev(8, 1), st_ino=768997, st_mode=S_IFREG|0644, st_nlink=1, st_uid=0, st_gid=0, st_blksize=4096, st_blocks=2128, st_size=1088952, st_atime=2019/03/21-18:43:37.128581999, st_mtime=2019/02/05-20:11:22, st_ctime=2019/03/21-18:43:36.208122000}) = 0
13924 mmap(NULL, 3178744, PROT_READ|PROT_EXEC, MAP_PRIVATE|MAP_DENYWRITE, 6</lib/x86_64-linux-gnu/libm-2.23.so>, 0) = 0x7fa37c0ea000
13924 mprotect(0x7fa37c1f2000, 2093056, PROT_NONE) = 0
13924 mmap(0x7fa37c3f1000, 8192, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 6</lib/x86_64-linux-gnu/libm-2.23.so>, 0x107000) = 0x7fa37c3f1000
13924 close(6</lib/x86_64-linux-gnu/libm-2.23.so>) = 0
13924 syscall_18446744073709551615(0x7ffe0f5a6d60, 0x80000, 0x7fa37efbd500, 0x7fa37edb56f0, 0, 0x7ffe0f5a6f7f) = -1 (errno 1)
13924 access("/etc/ld.so.nohwcap", F_OK) = -1 ENOENT (No such file or directory)
13924 syscall_18446744073709551615(0x7fa37efb5fd0, 0x80000, 0x7fa37efbd500, 0x6f732e6572637062, 0, 0x7ffe0f5a6f7f) = -1 (errno 1)
13924 syscall_18446744073709551615(0x7ffe0f5a6d60, 0x80000, 0x7fa37efbd500, 0x6f732e6572637062, 0, 0x7ffe0f5a6f7f) = -1 (errno 1)
13924 stat("/lib/x86_64-linux-gnu/tls/x86_64", 0x7ffe0f5a6e30) = -1 ENOENT (No such file or directory)
13924 syscall_18446744073709551615(0x7ffe0f5a6d60, 0x80000, 0x7fa37efbd500, 0x6f732e6572637062, 0, 0x7ffe0f5a6f7f) = -1 (errno 1)
13924 stat("/lib/x86_64-linux-gnu/tls", 0x7ffe0f5a6e30) = -1 ENOENT (No such file or directory)
13924 syscall_18446744073709551615(0x7ffe0f5a6d60, 0x80000, 0x7fa37efbd500, 0x6f732e6572637062, 0, 0x7ffe0f5a6f7f) = -1 (errno 1)
13924 stat("/lib/x86_64-linux-gnu/x86_64", 0x7ffe0f5a6e30) = -1 ENOENT (No such file or directory)
13924 syscall_18446744073709551615(0x7ffe0f5a6d60, 0x80000, 0x7fa37efbd500, 0x6f732e6572637062, 0, 0x7ffe0f5a6f7f) = -1 (errno 1)
13924 stat("/lib/x86_64-linux-gnu", {st_dev=makedev(8, 1), st_ino=768967, st_mode=S_IFDIR|0755, st_nlink=3, st_uid=0, st_gid=0, st_blksize=4096, st_blocks=8, st_size=4096, st_atime=2019/03/21-18:43:37.432733999, st_mtime=2019/03/21-18:43:36, st_ctime=2019/03/21-18:43:36.236136000}) = 0
13924 mmap(NULL, 4096, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7fa37efb4000
13924 writev(2</dev/null>, [{"/lib/systemd/systemd-journald", 29}, {": ", 2}, {"error while loading shared libra"..., 36}, {": ", 2}, {"libpcre.so.3", 12}, {": ", 2}, {"cannot open shared object file", 30}, {": ", 2}, {"Operation not permitted", 23}, {"\n", 1}], 10) = 139
13924 exit_group(127)                   = ?
```
I haven't tried to decipher it yet. Removing `RestrictSUIDSGID` "fixes" the issue.

**Comment by poettering** [comment]
> It looks like the `systemd-journald.service` fails because `systemd-journald-audit.socket` fails with `ConditionCapability=CAP_AUDIT_READ failed`, which, in turn, causes fail of `systemd-udevd-kernel.socket` and `systemd-udevd.service` and that causes the fail in mounting the rootfs, thus the emergency shell.

So, the systemd-journald-audit.socket condition fail is kinda expected (it is supposed to fail if audit privs are missing) and should not cause anything esle to fail. all it should do is a silent skipping of the audit hookup of journald if audit is not around.

**Comment by mrc0mmand** [comment]
Ah, I see, sorry for the confusion. However, the systemd-journald fail is relevant I guess:

```
[    3.270971] systemd[1]: Child 180 (systemd-journal) died (code=exited, status=127/n/a)
[    3.273899] systemd[1]: systemd-journald.service: Child 180 belongs to systemd-journald.service.
[    3.277108] systemd[1]: systemd-journald.service: Main process exited, code=exited, status=127/n/a
[    3.280623] systemd[1]: systemd-journald.service: Failed with result 'exit-code'.
[    3.283368] systemd[1]: systemd-journald.service: Changed start -> failed
[    3.293230] systemd[1]: systemd-journald.service: Job 21 systemd-journald.service/start finished, result=failed
```

**Comment by poettering** [comment]
I wonder what "syscall_18446744073709551615" is in the strace output @evverx pasted

**Comment by mrc0mmand** [comment]
This looks somewhat relevant:

> 
> 
> syscall_326 is `dup3` - from the `man` page: `dup3() was added to Linux in version 2.6.27; glibc support is available starting with version 2.9.`
> 
> Seems the `dup3` call isn't supported in the environment, and then something inside docker is getting confused (it's using syscall_INT_MAX`). I'm not sure what the right solution is here.



(Quote from https://github.com/rust-lang/rust/issues/51266#issuecomment-393731164)

**Comment by poettering** [comment]
@mrc0mmand hmm, from that comment this looks like some brokeness with old docker.

Might make sense to simply skip this setting inside of the testcases when old docker is used. What's the best approach to do that?

**Comment by mrc0mmand** [comment]
@poettering after playing around with the CentOS CI VM and removing the syscall filters (so I can run the systemd-journald under strace), it all came down to one error:

```
open("/usr/lib/systemd/librt.so.1", O_RDONLY|O_CLOEXEC) = -1 EPERM (Operation not permitted)
open("/lib64/tls/x86_64/librt.so.1", O_RDONLY|O_CLOEXEC) = -1 ENOSYS (Function not implemented)
stat("/lib64/tls/x86_64", 0x7ffce1886a60) = -1 EPERM (Operation not permitted)
stat("/lib64/tls", 0x7ffce1886a60)      = -1 ENOENT (No such file or directory)
open("/lib64/x86_64/librt.so.1", O_RDONLY|O_CLOEXEC) = -1 ENOSYS (Function not implemented)
stat("/lib64/x86_64", 0x7ffce1886a60)   = -1 EPERM (Operation not permitted)
stat("/lib64", {st_mode=S_IFDIR|0755, st_size=0, ...}) = 0
writev(2, [{"/usr/lib/systemd/systemd-journal"..., 33}, {": ", 2}, {"error while loading shared libra"..., 36}, {": ", 2}, {"librt.so.1", 10}, {": ", 2}, {"cannot open shared object file", 30}, {": ", 2}, {"Operation not permitted", 23}, 1
exit_group(127)                         = ?
+++ exited with 127 +++

```

**Comment by mrc0mmand** [comment]
Just for a reference, here's a full trace taken from the dracut emergency shell:

```
dracut:/# cat strace-service.log.3026
execve("/usr/lib/systemd/systemd-journald", ["/usr/lib/systemd/systemd-journal"...], [/* 9 vars */]) = 0
brk(NULL)                               = 0xb47000
mmap(NULL, 4096, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f4e20555000
access("/etc/ld.so.preload", R_OK)      = -1 ENOENT (No such file or directory)
open("/usr/lib/systemd/tls/x86_64/libsystemd-shared-241.so", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
stat("/usr/lib/systemd/tls/x86_64", 0x7ffce1886cc0) = -1 ENOENT (No such file or directory)
open("/usr/lib/systemd/tls/libsystemd-shared-241.so", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
stat("/usr/lib/systemd/tls", 0x7ffce1886cc0) = -1 ENOENT (No such file or directory)
open("/usr/lib/systemd/x86_64/libsystemd-shared-241.so", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
stat("/usr/lib/systemd/x86_64", 0x7ffce1886cc0) = -1 ENOENT (No such file or directory)
open("/usr/lib/systemd/libsystemd-shared-241.so", O_RDONLY|O_CLOEXEC) = 6
read(6, "\177ELF\2\1\1\0\0\0\0\0\0\0\0\0\3\0>\0\1\0\0\0\20\324\4\0\0\0\0\0"..., 832) = 832
fstat(6, {st_mode=S_IFREG|0755, st_size=3808936, ...}) = 0
mmap(NULL, 5496304, PROT_READ|PROT_EXEC, MAP_PRIVATE|MAP_DENYWRITE, 6, 0) = 0x7f4e1fdf7000
mprotect(0x7f4e200a5000, 2097152, PROT_NONE) = 0
mmap(0x7f4e202a5000, 589824, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 6, 0x2ae000) = 0x7f4e202a5000
close(6)                                = 0
open("/usr/lib/systemd/libselinux.so.1", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
open("/etc/ld.so.cache", O_RDONLY|O_CLOEXEC) = 6
fstat(6, {st_mode=S_IFREG|0644, st_size=10026, ...}) = 0
mmap(NULL, 10026, PROT_READ, MAP_PRIVATE, 6, 0) = 0x7f4e20552000
close(6)                                = 0
open("/lib64/libselinux.so.1", O_RDONLY|O_CLOEXEC) = 6
read(6, "\177ELF\2\1\1\0\0\0\0\0\0\0\0\0\3\0>\0\1\0\0\0\320i\0\0\0\0\0\0"..., 832) = 832
fstat(6, {st_mode=S_IFREG|0755, st_size=155784, ...}) = 0
mmap(NULL, 4096, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f4e20551000
mmap(NULL, 2255184, PROT_READ|PROT_EXEC, MAP_PRIVATE|MAP_DENYWRITE, 6, 0) = 0x7f4e1fbd0000
mprotect(0x7f4e1fbf4000, 2093056, PROT_NONE) = 0
mmap(0x7f4e1fdf3000, 8192, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 6, 0x23000) = 0x7f4e1fdf3000
mmap(0x7f4e1fdf5000, 6480, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_ANONYMOUS, -1, 0) = 0x7f4e1fdf5000
close(6)                                = 0
open("/usr/lib/systemd/libpthread.so.0", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
open("/lib64/libpthread.so.0", O_RDONLY|O_CLOEXEC) = 6
read(6, "\177ELF\2\1\1\0\0\0\0\0\0\0\0\0\3\0>\0\1\0\0\0\260l\0\0\0\0\0\0"..., 832) = 832
fstat(6, {st_mode=S_IFREG|0755, st_size=134608, ...}) = 0
mmap(NULL, 2208904, PROT_READ|PROT_EXEC, MAP_PRIVATE|MAP_DENYWRITE, 6, 0) = 0x7f4e1f9b4000
mprotect(0x7f4e1f9cb000, 2093056, PROT_NONE) = 0
mmap(0x7f4e1fbca000, 8192, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 6, 0x16000) = 0x7f4e1fbca000
mmap(0x7f4e1fbcc000, 13448, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_ANONYMOUS, -1, 0) = 0x7f4e1fbcc000
close(6)                                = 0
open("/usr/lib/systemd/libc.so.6", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
open("/lib64/libc.so.6", O_RDONLY|O_CLOEXEC) = 6
read(6, "\177ELF\2\1\1\3\0\0\0\0\0\0\0\0\3\0>\0\1\0\0\0\340$\2\0\0\0\0\0"..., 832) = 832
fstat(6, {st_mode=S_IFREG|0755, st_size=2107544, ...}) = 0
mmap(NULL, 3981792, PROT_READ|PROT_EXEC, MAP_PRIVATE|MAP_DENYWRITE, 6, 0) = 0x7f4e1f5e7000
mprotect(0x7f4e1f7a9000, 2097152, PROT_NONE) = 0
mmap(0x7f4e1f9a9000, 24576, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 6, 0x1c2000) = 0x7f4e1f9a9000
mmap(0x7f4e1f9af000, 16864, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_ANONYMOUS, -1, 0) = 0x7f4e1f9af000
close(6)                                = 0
open("/usr/lib/systemd/librt.so.1", O_RDONLY|O_CLOEXEC) = -1 EPERM (Operation not permitted)
open("/lib64/tls/x86_64/librt.so.1", O_RDONLY|O_CLOEXEC) = -1 ENOSYS (Function not implemented)
stat("/lib64/tls/x86_64", 0x7ffce1886a60) = -1 EPERM (Operation not permitted)
stat("/lib64/tls", 0x7ffce1886a60)      = -1 ENOENT (No such file or directory)
open("/lib64/x86_64/librt.so.1", O_RDONLY|O_CLOEXEC) = -1 ENOSYS (Function not implemented)
stat("/lib64/x86_64", 0x7ffce1886a60)   = -1 EPERM (Operation not permitted)
stat("/lib64", {st_mode=S_IFDIR|0755, st_size=0, ...}) = 0
writev(2, [{"/usr/lib/systemd/systemd-journal"..., 33}, {": ", 2}, {"error while loading shared libra"..., 36}, {": ", 2}, {"librt.so.1", 10}, {": ", 2}, {"cannot open shared object file", 30}, {": ", 2}, {"Operation not permitted", 23}, 1
exit_group(127)                         = ?
+++ exited with 127 +++
```

**Comment by mrc0mmand** [comment]
> @mrc0mmand hmm, from that comment this looks like some brokeness with old docker.
> 
> Might make sense to simply skip this setting inside of the testcases when old docker is used. What's the best approach to do that?

It indeed looks like some weirdness in docker, but given the strace above the similar issue (with missing permissions for loading of a shared library) happens in CentOS CI under a KVM as well.

**Comment by mrc0mmand** [comment]
Also, the offending library (`librt.so`) is installed on the system, however, under `/usr/lib64/librt.so`
```
dracut:/# ls -l /usr/lib64/librt*  
-rwxr-xr-x 1 root root 42512 Mar 22 12:03 /usr/lib64/librt-2.17.so
lrwxrwxrwx 1 root root    13 Mar 22 12:03 /usr/lib64/librt.so -> librt-2.17.so
lrwxrwxrwx 1 root root    13 Mar 22 12:03 /usr/lib64/librt.so.1 -> librt-2.17.so
```

This path is not even checked according to the strace above.

**Comment by evverx** [comment]
> Might make sense to simply skip this setting inside of the testcases when old docker is used. 

Just to clarify, trying to reproduce the issue I didn't use docker at all. It was `journald` in an nspawn container on top of Ubuntu Xenial.

Given that the setting clearly breaks something at least on CentOS CI and  Ubuntu Xenial (effectively rendering them nonoperational), I think it'd be better to figure out what is going on instead of skipping the tests. It's true that in, for example, https://github.com/systemd/systemd/issues/9666 it was decided that the test should be skipped but in that case the filters simply don't work, which isn't as bad as all the basic services suddenly failing to start sending their logs to `/dev/null` along the way.

**Comment by poettering** [comment]
@evverx the problem is that this is unlikely something we can fix in systemd. I mean, the seccomp filter this PR installs is simple. It just says say EPERM if the third argument to open() matches something. What the straces show however is that the whole syscall gets entirely corrupted (and ENOSYS is returned). THis is almost definitely a problem between the libseccomp used and the older kernel used on those old systems, and I am not a libseccomp hacker to do much about it, and I doubt many people really want to debug this on old OSes.

Or in other words: it's totally worth it supporting systemd on various systems, including older ones, and fixing issues we find in systemd as they pop up then. But as soon as we come into territory where bugs are between two unrelated non-systemd packages and we are just triggering them we have a problem, if these are not issues we have the expertise or interest to debug.

At least I know for sure my interest in tracking down libseccomp issues on old CentOS is relatively minimal, and I don't see anyone else willing to do that either. Hence I propose just skipping this specific test on such combinations.

**Comment by poettering** [comment]
(it might be worth checking if updating libseccomp and/or the kernel on the CI systems to something current will make the issue go away)

**Comment by mrc0mmand** [comment]
Given the recompilation of `libseccomp-2.4.0` in el7 buildroot went flawlessly (the current CentOS 7 version is `2.3.1`), I can definitely try it and see if it fixes the issue.

**Comment by mrc0mmand** [comment]
Newer `libseccomp` was apparently not enough, so I tried also a newer kernel (`kernel-lt-4.4.177-1.el7` from the [ELRepo](https://elrepo.org/tiki/tiki-index.php) project)

```
# rpm --import https://www.elrepo.org/RPM-GPG-KEY-elrepo.org
# yum -y install https://www.elrepo.org/elrepo-release-7.0-3.el7.elrepo.noarch.rpm
# yum -y install --enablerepo=elrepo-kernel kernel-lt
# dracut -f --kver "$(rpm -q --qf "%{VERSION}-%{RELEASE}.%{ARCH}" kernel-lt)"
# grub2-mkconfig -o /boot/grub2/grub.cfg
# sed -i'' '/saved_entry/d' /boot/grub2/grubenv
# systemctl reboot
...
# uname -r
4.4.177-1.el7.elrepo.x86_64
...
# ninja-build -C build
...
Dependency libseccomp found: YES 2.4.0
...
```

The *workaround* above is somewhat ugly and some of the steps are probably unnecessary. Nevertheless, the issue is still present:

```
...
[    0.000000] Linux version 4.4.177-1.el7.elrepo.x86_64 (mockbuild@Build64R7) (gcc version 4.8.5 20150623 (Red Hat 4.8.5-36) (GCC) )9
...
[   31.479839] systemd[1]: Received SIGCHLD from PID 213 (systemd-journal).
[   31.515183] systemd[1]: Child 213 (systemd-journal) died (code=exited, status=127/n/a)
[   31.585112] systemd[1]: systemd-journald.service: Child 213 belongs to systemd-journald.service.
[   31.597931] systemd[1]: systemd-journald.service: Main process exited, code=exited, status=127/n/a
[   31.655868] systemd[1]: systemd-journald.service: Failed with result 'exit-code'.
[   31.660070] systemd[1]: systemd-journald.service: Changed start -> failed
[   31.829008] systemd[1]: systemd-journald.service: Job 51 systemd-journald.service/start finished, result=failed
[   31.838341] systemd[1]: Failed to start Journal Service.
[FAILED] Failed to start Journal Service.
See 'systemctl status systemd-journald.service' for details.
```

**Comment by mrc0mmand** [comment]
Just out of curiosity I tried it with the mainline kernel version (`kernel-ml-5.0.4-1.el7.elrepo.x86_64`).

```
[    0.000000] Linux version 5.0.4-1.el7.elrepo.x86_64 (mockbuild@Build64R7) (gcc version 4.8.5 20150623 (Red Hat 4.8.5-36) (GCC)) #19
....
[   20.822811] systemd[1]: Received SIGCHLD from PID 220 (systemd-journal).
[   20.825298] systemd[1]: Child 220 (systemd-journal) died (code=exited, status=127/n/a)
[   20.828904] systemd[1]: systemd-journald.service: Child 220 belongs to systemd-journald.service.
[   20.830060] systemd[1]: systemd-journald.service: Main process exited, code=exited, status=127/n/a
[   20.837574] systemd[1]: systemd-journald.service: Failed with result 'exit-code'.
[   20.838219] systemd[1]: systemd-journald.service: Changed start -> failed
[   20.855191] systemd[1]: systemd-journald.service: Job 12 systemd-journald.service/start finished, result=failed
[   20.855907] systemd[1]: Failed to start Journal Service.
[FAILED] Failed to start Journal Service.
```

I guess there's some other dependency, which has to be upgraded as well...

**Comment by poettering** [comment]
hmm, recent glibc's use openat() preferably over open() and internally direct the glibc function open() to use the openat() syscall internally (with the first param set to AT_FDCWD). I wonder if this might be triggering the problem: i..e is it possible this goes away when a new glibc is used that uses openat() instead of open() to implement open()?

**Comment by poettering** [comment]
hmm, i have a suspicion what might be causing this, lemme see.

**Comment by mrc0mmand** [comment]
```
TEST RUN: Basic systemd setup [OK]
TEST CLEANUP: Basic systemd setup
```

Nice!

**Comment by poettering** [comment]
Neat. Seems to have passed now. So I don't fully grok it, but here's what I changed:

open()/openat() are a bit weird. They last argument the call gets (i.e. the mode_t one) is optional, and only has to be supplied if O_CREAT is specified. The original PR didn't care for the fact it's optional and always checked the last argument. glibc actually passes rubbish as last argument to the kernel when it's not specified, and hence we tested this rubbish. In the newest version i check if O_CREAT is set before looking at the last argument. This apparently fixed the issue.

I am not entirely sure what's going on though, since the effect of this missing check should have been that we randomly somtimes would have returned EPERM when O_CREAT wasn't set. However, what actually happened was that ENOSYS was returned, which I don#t really grrok (maybe the seccomp kernel side actually notices when the seccomp filter checks for more args than passed? dunno…). 

Either way, this definitely was triggered by a bug in my code. Hence apologies for all the confusion I created. Sorry!

PTAL!

**Comment by ret2libc** [comment]
Nice catch!

File: man/systemd.exec.xml [hunk 2]
```diff
@@ -1392,6 +1393,21 @@ RestrictNamespaces=~cgroup net</programlisting>
         that actually require them. Defaults to off.</para></listitem>
       </varlistentry>
 
+      <varlistentry>
+        <term><varname>RestrictSUIDSGID=</varname></term>
+
+        <listitem><para>Takes a boolean argument. If set, any attempts to set the set-user-ID (SUID) or
+        set-group-ID (SGID) bits on files or directories will be denied (for details on these bits see
+        <citerefentry><refentrytitle>inode</refentrytitle><manvolnum>7</manvolnum></citerefentry>). If
```
Comment (keszybz) [inline man/systemd.exec.xml]: This needs a `project=...` attribute so the href is not broken.

Comment (poettering) [reply man/systemd.exec.xml]: will fix

**Comment by poettering** [comment]
Force pushed a new version. Only change is the man page fix.

**Comment by poettering** [comment]
Taking the liberty to set green label since so many folks already liked it.

**Pull request merged**